### PR TITLE
CAMEL-15337: create toc for hugo based content

### PR DESCRIPTION
**Pull request by AemieJ**
### Motive for building Hugo toc
For the apache camel website, the antora pages have a toc generated which will help in easing the user navigation. Thus, for the Hugo pages, toc should be implemented. 

### DIfference between toc of Antora and Hugo 
The manner in which the toc of Antora and Hugo is structured is completed different. To trigger the toc for Hugo, one code of line needs to be inserted i.e `{{ TableOfContents }}` however I wanted the UI implementation of both Antora and Hugo to be the same.
In Hugo, the manner in which toc is created is just a nested loop of `ul` and `li` but in the case of Antora, it's a proper structure with defined levels where `##`  would signify `data-level: 2` and `###` would signify `data-level:3`. This wasn't the case of Hugo. 

Thus, my first implementation within the **js** file was to create the data levels.

### Problems faced while implementing fragment jumping of links and also based on scrolling. 
The `scroll` event function that is implemented for antora in the `on-this-page.js` can't be used for the Hugo as their structures are completely different as stated before. Hence, I have implemented my own logic similar to that of antora for the scroll and fragment jumping and it works. 

#### Drawbacks to my implementation currently

In my implementation, the fragment jumping works smoothly when done through the keyboard however when using a scrollbar, it shows odd behavior on scrolling down in some cases. 

toc-content-hugo

### CODE REVIEW

File: content/community/articles.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 title: "Articles"
 ---
 
-### Camel Videos
+## Camel Videos
```
Comment (zregvart) [inline content/community/articles.md:5]: These changes could be on another PR, they don't seem to be related to this.

File: antora-ui-camel/src/css/main.css [hunk 2]
```diff
@@ -1,5 +1,6 @@
 @media screen and (max-width: 1023px) {
-  aside.toc.sidebar {
+  aside.toc.sidebar,
+  aside.toc-sidebar {
```
[the remark this answers is not in the document]
Comment (AemieJ) [reply antora-ui-camel/src/css/main.css:3]: I need to currently find the issue 'cause whenever I use `<aside class="toc sidebar">` in case of Hugo content, this element doesn't render on the page. That's why I temporarily made use of `<aside class="toc toc-sidebar">`

File: antora-ui-camel/src/css/static.css [hunk 3]
```diff
@@ -1,6 +1,11 @@
 .static {
   margin: var(--static-margin);
-  max-width: var(--static-max-width--desktop);
+}
+
+@media screen and (min-width: 1024px) {
+  .static {
+    max-width: var(--static-max-width--desktop);
+  }
```
[the remark this answers is not in the document]
Comment (AemieJ) [reply antora-ui-camel/src/css/static.css:8]: Each change in this file is related to the toc. The thing is within the antora pages fragment jumping occurs accurately with respect to the toolbar. As I didn't want to include additional code for fragment jumping and positioning, I included a blank toolbar in the case of Hugo pages as well.

File: antora-ui-camel/src/css/main.css [hunk 4]
```diff
@@ -28,7 +30,8 @@
     scrollbar-width: thin; /* Firefox */
   }
 
-  .toc.sidebar .toc-menu ul::-webkit-scrollbar {
+  .toc.sidebar .toc-menu ul::-webkit-scrollbar,
+  .toc-sidebar .toc-hugo ul::-webkit-scrollbar {
```
[the remark this answers is not in the document]
Comment (AemieJ) [reply antora-ui-camel/src/css/main.css:34]: I didn't exactly get what do you mean by this?

File: antora-ui-camel/src/js/98-hugo-on-this-page.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+  var child = []
+  var lastActiveLink = null
+  var selector = document.querySelector('.toolbar')
+  var headings = Array.from(document.querySelectorAll('article.doc h1, article.doc h2, article.doc h3'))
+  var links = Array.from(document.querySelectorAll('aside.toc-sidebar .toc-hugo a'))
+  var width = document.querySelector('.navbar').offsetWidth
+  var parentSelector = document.querySelectorAll('.toc-hugo li ul:first-child')
+
+  var topValMin = selector.getBoundingClientRect().top + selector.getBoundingClientRect().height - 5
+  var topValMax = topValMin + 20
```
[the remark this answers is not in the document]
Comment (AemieJ) [reply antora-ui-camel/src/js/98-hugo-on-this-page.js:14]: I have worked with changing the font size in the browser settings, it works as this just sets the range based on the position and height of the toolbar.

File: antora-ui-camel/src/css/toolbar.css [hunk 6]
```diff
@@ -12,6 +12,10 @@
   z-index: var(--z-index-toolbar);
 }
 
+.toolbar.hugo {
+  box-shadow: none;
+}
```
[the remark this answers is not in the document]
Comment (AemieJ) [reply antora-ui-camel/src/css/toolbar.css:17]: The thing is within the antora pages fragment jumping occurs accurately with respect to the toolbar. As I didn't want to include additional code for fragment jumping and positioning, I included a blank toolbar in the case of Hugo pages as well.

File: content/community/articles.md (hunk 1 above)
Comment (AemieJ) [reply content/community/articles.md:5]: Well, a few of the contents within the Hugo pages weren't perfectly aligned due to which the toc content wasn't being generated accurately. Hence, I had to make changes to these particular files.

**Comment by AemieJ** [comment]
@zregvart I did research on **Intersection Observer** and tried its implementation but on observing, it is only possible with a parent and child element. Here, `article.doc` is the parent ( root ) element however it doesn't cover a certain height thus the visibility of the target ( all the headings ) can't be compared based on it. Also, it has a loophole, the threshold triggers for any of its headings on the page.

Other than that, I modified the implementation of logic, currently, it works pretty efficiently ( even when the font-size within browser settings is modified )

**Comment by zregvart** [comment]
With #459 merged I think we can close this now. @AemieJ is there anything I've missed?
